### PR TITLE
fix(comment-url): hold fragment if there's any

### PR DESCRIPTION
**Pull request by tx0c**
1. for comment url https://web-next.matters.news/p/30000#comments
2. for particular comment url https://web-next.matters.news/@smog_again/30000#Q29tbWVudDoyNzg5Njc-Q29tbWVudDoyNzkwMDY
3. for article owner's `?mode=edit` to enter edit mode

also add utm tracking code, for better computing links source&destination before or after visiting a link

should also resolve https://github.com/thematters/matters-web/issues/2587

### CODE REVIEW

File: src/components/ClientUpdater/index.tsx [hunk 1]
```diff
@@ -47,14 +47,20 @@ export const ClientUpdater = () => {
...
+  ) => {
+    if (!client?.writeData || shallow) {
+      // skip shallow route change
       return
     }
 
     const newRouteHistory = [...routeHistoryRef.current, url]
     routeHistoryRef.current = newRouteHistory
 
+    console.log('update routeHistory to:', { url, newRouteHistory })
```
Comment (robertu7) [inline src/components/ClientUpdater/index.tsx:62]: 👀

File: src/components/Layout/Header/BackButton/index.tsx [hunk 2]
```diff
@@ -43,6 +43,8 @@ export const BackButton: React.FC<BackButtonProps> = ({
 
     const routeHistory = data?.clientPreference.routeHistory || []
 
+    console.log('onBack:', { routeHistory })
```
Comment (robertu7) [inline src/components/Layout/Header/BackButton/index.tsx:46]: 👀

File: src/views/ArticleDetail/index.tsx [hunk 3]
```diff
@@ -200,8 +200,32 @@ const ArticleDetail = () => {
...
+    // copy all others
+    const rems = [
+      ...u.searchParams, // uses .entries()
+      ...n.searchParams,
+    ].filter(([k, v]) => !k?.startsWith('utm_'))
+    const nsearch = rems.length > 0 ? `?${new URLSearchParams(rems)}` : ''
+    const nhref = `${n.pathname}${nsearch}${n.hash || u.hash}`
+
+    if (nhref !== router.asPath) {
+      console.log('replacing url:', {
```
Comment (robertu7) [inline src/views/ArticleDetail/index.tsx:222]: 👀

File: src/views/Home/Feed/MainFeed/index.tsx [hunk 4]
```diff
@@ -209,6 +209,7 @@ const MainFeed = ({ feedSortType: sortBy }: MainFeedProps) => {
             <List.Item key={`${sortBy}:${edge.node.id}`}>
               <ArticleDigestFeed
                 article={edge.node}
+                utm_source="homepage"
```
Comment (robertu7) [inline src/views/Home/Feed/MainFeed/index.tsx:212]: camelCase?

**Comment by tx0c** [comment]
should be already in 86a446a0d ; without debugging log lines

**Pull request closed without merging**